wdl/ExpansionHunterDenovo.wdl

```diff
@@ -0,0 +1,347 @@
+## ExpansionHunter denovo (EHdn)
+##
+## This WDL implements workflows for
+## EHdn's two execution modes:
+## (i) case-control analysis;
+## (ii) outliner analysis.
+
+version 1.0
+
+import "Structs.wdl"
+
+struct FilenamePostfixes {
+  String locus
+  String motif
+  String profile
+  String merged_profile
+  Int profile_len
+}
+
+workflow EHdnSTRAnalysis {
+
+  input {
+    String analysis_type
+    String str_comparison_type
+    Array[File] case_reads_filenames
+    Array[File] case_indexes_filenames
+    Array[File] control_reads_filenames
+    Array[File] control_indexes_filenames
```

Comment:
Does ehdn support bams or crams or both? I would use that in the variable name e.g. `case_crams` or `case_crams_or_bams`.

Reply:
Also this will be tricky to map to the Terra data model. I would propose having a single array for the reads and a single array for the indexes (without case/control designation), and then have a third array `Array[String] sample_status` with possible values `case` or `control`. Then you can use a scatter to select each like this:
```
scatter (i in range(length(sample_crams_or_bams)) {
    if (sample_status[i] == "case") {
        # Creates Array[File?] case_crams_or_bams_ and case_indexes_
        File case_crams_or_bams_ = sample_crams_or_bams[i]
        File case_indexes_ = sample_indexes[i]
    }
    if (sample_status[i] == "control") {
        File control_crams_or_bams_ = sample_crams_or_bams[i]
        File control_indexes_ = sample_indexes[i]
    }
}
```
and then when calling the Case task you just use `select_all` like this:
```
reads_filename = select_all(case_crams_or_bams_),
index_filename = select_all(case_indexes_)
```
and similarly for the Controls call.

Reply:
ExpansionHunter denovo requires a [manifest file](https://github.com/Illumina/ExpansionHunterDenovo/blob/master/examples/case-control/manifest.tsv) that serves a similar purpose to the `sample_status` array you suggested. I found compiling that file would be a bit cumbersome for user and error prune so instead, I switched to using two lists and auto-generating the manifest accordingly. However, if tricky on terra, I'm happy to revert it. Which one do you think will be more intuitive and simpler on Terra, manifest file or `sample_status` array?

Reply:
Terra has users upload sample-level metadata as a table and define sets of samples (batches) using sample IDs. The user can then launch batched workflows, like this one, with some inputs (eg sample id, cram path, or case/control status) generated from columns of the table as arrays. 

They would have to create and upload the manifest file separately and then manually enter the bucket URL into the inputs. You're right - that's extra work and prone to errors. Therefore we should use arrays for all the inputs and compose the manifest on the fly, even if it's a small cost. 

Note also that Terra does not support certain WDL functions so you may need to write a task to create the manifest (see [here](https://support.terra.bio/hc/en-us/community/posts/360071465631-write-lines-write-map-write-tsv-write-json-fail-when-run-in-a-workflow-rather-than-in-a-task)).